File: uni_modules/tuikit-atomic-x/utssdk/app-ios/swift/AlbumPickerManager.swift
import AlbumPicker
import DCloudUTSFoundation
import UIKit

private let APM_TAG: String = "iOS-AlbumPickerManager: "

private struct SessionCallbacks {
    let onPickConfirm: (String) -> Void
    let onMediaProcessing: (String) -> Void
    let onMediaProcessed: () -> Void
    let onCancel: () -> Void
}

public class AlbumPickerManager: NSObject {

    public static let shared = AlbumPickerManager()

    private var hostVC: UIViewController?
    private var pickerView: AlbumPicker.AlbumPickerView?
    private var sessionCallbacks: [String: SessionCallbacks] = [:]
    private var delegateProxies: [String: AlbumPickerDelegateProxy] = [:]
    private var currentSessionId: String?

    private override init() {
        super.init()
    }

    // MARK: - Public API

    public func show(
        _ configJSON: String,
        _ themeJSON: String,
        sessionId: String,
        viewController: UIViewController,
        onPickConfirm: @escaping (String) -> Void,
        onMediaProcessing: @escaping (String) -> Void,
        onMediaProcessed: @escaping () -> Void,
        onCancel: @escaping () -> Void
    ) {
        console.log(APM_TAG, "show called with sessionId:", sessionId)

        sessionCallbacks[sessionId] = SessionCallbacks(
            onPickConfirm: onPickConfirm,
            onMediaProcessing: onMediaProcessing,
            onMediaProcessed: onMediaProcessed,
            onCancel: onCancel
        )
        currentSessionId = sessionId

        // Dismiss existing if any
        if let existingVC = hostVC {
            existingVC.dismiss(animated: false)
            self.hostVC = nil
            self.pickerView = nil
        }

        let vc = UIViewController()
        vc.modalPresentationStyle = .fullScreen
        vc.view.backgroundColor = .black

        let config = parseConfig(configJSON)
        let theme = parseTheme(themeJSON)

        let picker = AlbumPicker.AlbumPickerView(frame: vc.view.bounds)

        let proxy = AlbumPickerDelegateProxy(sessionId: sessionId, manager: self)
        delegateProxies[sessionId] = proxy
        picker.delegate = proxy

        picker.initialize(config: config, theme: theme)

        picker.translatesAutoresizingMaskIntoConstraints = false
        vc.view.addSubview(picker)

        NSLayoutConstraint.activate([
            picker.leadingAnchor.constraint(equalTo: vc.view.leadingAnchor),
            picker.trailingAnchor.constraint(equalTo: vc.view.trailingAnchor),
            picker.topAnchor.constraint(equalTo: vc.view.topAnchor),
            picker.bottomAnchor.constraint(equalTo: vc.view.bottomAnchor)
        ])

        self.pickerView = picker
        self.hostVC = vc

        if viewController.presentedViewController != nil {
            viewController.dismiss(animated: false) {
                viewController.present(vc, animated: true)
            }
        } else {
            viewController.present(vc, animated: true)
        }
        console.log(APM_TAG, "show: presented picker VC, sessionId:", sessionId)
    }

    public func hide() {
        console.log(APM_TAG, "hide called")
        guard let vc = hostVC else {
            console.warn(APM_TAG, "hide: no active picker to dismiss")
            return
        }
        vc.dismiss(animated: true) { [weak self] in
            self?.hostVC = nil
            console.log(APM_TAG, "cleanupView done")
        }
    }

    fileprivate func getCallbacks(sessionId: String) -> SessionCallbacks? {
        return sessionCallbacks[sessionId]
    }

    fileprivate func cleanup(sessionId: String) {
        sessionCallbacks.removeValue(forKey: sessionId)
        delegateProxies.removeValue(forKey: sessionId)
        if currentSessionId == sessionId {
            pickerView?.removeFromSuperview()
            pickerView = nil
            hostVC = nil
            currentSessionId = nil
        }
        console.log(APM_TAG, "cleanup done for sessionId:", sessionId)
    }

    fileprivate func dismissHost(sessionId: String) {
        hostVC?.dismiss(animated: true) { [weak self] in
            self?.cleanup(sessionId: sessionId)
        }
    }

    // MARK: - JSON Parsing

    private func parseConfig(_ jsonStr: String) -> AlbumPicker.AlbumPickerConfig {
        var config = AlbumPicker.AlbumPickerConfig()

        guard let data = jsonStr.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return config
        }

        if let pickMode = dict["pickMode"] as? Int {
            switch pickMode {
            case 0: config.mediaFilter = .imageOnly
            case 1: config.mediaFilter = .videoOnly
            case 2: config.mediaFilter = .imageAndVideo
            default: break
            }
        }

        if let maxCount = dict["maxCount"] as? Int { config.maxSelectionCount = maxCount }
        if let gridCount = dict["gridCount"] as? Int { config.itemsPerRow = gridCount }
        if let showsCamera = dict["showsCameraItem"] as? Bool { config.showsCameraItem = showsCamera }

        if let style = dict["style"] as? Int {
            switch style {
            case 0: config.style = .likeWeChat
            case 1: config.style = .likeWhatsApp
            default: break
            }
        }

        if let language = dict["language"] as? Int {
            switch language {
            case 0: config.language = .system
            case 1: config.language = .en
            case 2: config.language = .zhHans
            case 3: config.language = .zhHant
            case 4: config.language = .ar
            default: break
            }
        }

        if let compressQuality = dict["compressQuality"] as? Int {
            switch compressQuality {
            case 0: config.compressQuality = .standard
            case 1: config.compressQuality = .high
            default: break
            }
        }

        if let maxVideoDuration = dict["maxVideoDurationInSeconds"] as? Int {
            config.maxVideoDurationInSeconds = maxVideoDuration
        }
        if let maxOutputFileSize = dict["maxOutputFileSizeInMB"] as? Int {
            config.maxOutputFileSizeInMB = maxOutputFileSize
        }

        return config
    }

    private func parseTheme(_ jsonStr: String) -> AlbumPicker.AlbumPickerTheme {
        var theme = AlbumPicker.AlbumPickerTheme()

        guard let data = jsonStr.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return theme
        }

        theme.currentPrimaryColor = parseColor(dict["primaryColor"])
        theme.backgroundColor = parseColor(dict["backgroundColor"])
        theme.backgroundColorSecondary = parseColor(dict["backgroundColorSecondary"])
        theme.textColor = parseColor(dict["textColor"])
        theme.textColorSecondary = parseColor(dict["textColorSecondary"])

        if let v = dict["bigFontSize"] as? Double { theme.bigFontSize = CGFloat(v) }
        if let v = dict["normalFontSize"] as? Double { theme.normalFontSize = CGFloat(v) }
        if let v = dict["smallFontSize"] as? Double { theme.smallFontSize = CGFloat(v) }
        if let v = dict["bigRadius"] as? Double { theme.bigRadius = CGFloat(v) }
        if let v = dict["normalRadius"] as? Double { theme.normalRadius = CGFloat(v) }
        if let v = dict["smallRadius"] as? Double { theme.smallRadius = CGFloat(v) }

        return theme
    }

    private func parseColor(_ value: Any?) -> UIColor? {
        guard let hexStr = value as? String else { return nil }
        let hex = hexStr.replacingOccurrences(of: "0x", with: "")
                        .replacingOccurrences(of: "0X", with: "")
                        .replacingOccurrences(of: "#", with: "")
        guard hex.count == 8, let intVal = UInt64(hex, radix: 16) else { return nil }

        let a = CGFloat((intVal >> 24) & 0xFF) / 255.0
        let r = CGFloat((intVal >> 16) & 0xFF) / 255.0
        let g = CGFloat((intVal >> 8) & 0xFF) / 255.0
        let b = CGFloat(intVal & 0xFF) / 255.0
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }

    fileprivate func serializeAlbumMedia(_ media: AlbumPicker.AlbumMedia) -> [String: Any] {
        var dict: [String: Any] = [
            "id": media.id,
            "mediaType": media.mediaType == .video ? 1 : 0,
            "mediaPath": media.mediaPath ?? "",
            "fileExtension": "",
            "fileSize": 0,
            "duration": media.duration,
        ]
        if let thumbnail = media.videoThumbnailPath {
            dict["videoThumbnailPath"] = thumbnail
        }
        return dict
    }

    fileprivate func toJSONString(_ obj: Any) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: obj),
              let str = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return str
    }
}

// MARK: - AlbumPickerDelegateProxy

class AlbumPickerDelegateProxy: NSObject, AlbumPicker.AlbumPickerDelegate {
    let sessionId: String
    private weak var manager: AlbumPickerManager?

    init(sessionId: String, manager: AlbumPickerManager) {
        self.sessionId = sessionId
        self.manager = manager
        super.init()
    }

    public func onPickConfirm(pickedAlbumMedias: [AlbumPicker.AlbumMedia], textMessage: String?) {
        console.log(APM_TAG, "onPickConfirm, count:", pickedAlbumMedias.count, ", sessionId:", sessionId)

        guard let manager = manager, let callbacks = manager.getCallbacks(sessionId: sessionId) else {
            console.error(APM_TAG, "onPickConfirm: callbacks not found for sessionId:", sessionId)
            return
        }

        manager.hide()

        let serialized = pickedAlbumMedias.map { manager.serializeAlbumMedia($0) }
        let payload: [String: Any] = [
            "pickedAlbumMedias": serialized,
            "textMessage": textMessage ?? ""
        ]
        callbacks.onPickConfirm(manager.toJSONString(payload))
    }

    public func onMediaProcessing(albumMedia: AlbumPicker.AlbumMedia, progress: Float, error: Bool) {

        guard let manager = manager, let callbacks = manager.getCallbacks(sessionId: sessionId) else {
            console.error(APM_TAG, "onMediaProcessing: callbacks not found for sessionId:", sessionId)
            return
        }

        let payload: [String: Any] = [
            "albumMedia": manager.serializeAlbumMedia(albumMedia),
            "progress": Double(progress),
            "error": error
        ]
        callbacks.onMediaProcessing(manager.toJSONString(payload))
    }

    public func onMediaProcessed() {
        console.log(APM_TAG, "onMediaProcessed, sessionId:", sessionId)

        guard let manager = manager, let callbacks = manager.getCallbacks(sessionId: sessionId) else {
            console.error(APM_TAG, "onMediaProcessed: callbacks not found for sessionId:", sessionId)
            return
        }

        callbacks.onMediaProcessed()
        manager.cleanup(sessionId: sessionId)
    }

    public func onCancel() {
        console.log(APM_TAG, "onCancel, sessionId:", sessionId)

        guard let manager = manager, let callbacks = manager.getCallbacks(sessionId: sessionId) else {
            console.error(APM_TAG, "onCancel: callbacks not found for sessionId:", sessionId)
            return
        }

        manager.dismissHost(sessionId: sessionId)
        callbacks.onCancel()
    }
}

File: uni_modules/tuikit-atomic-x/utssdk/app-android/kotlin/AlbumPickerManager.kt
package uts.sdk.modules.atomicx.kotlin

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.app.Activity
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.FrameLayout
import androidx.activity.ComponentActivity
import androidx.lifecycle.findViewTreeLifecycleOwner
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.savedstate.findViewTreeSavedStateRegistryOwner
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import com.google.gson.Gson
import io.dcloud.uts.console
import io.trtc.tuikit.atomicx.albumpicker.AlbumMedia
import io.trtc.tuikit.atomicx.albumpicker.AlbumMediaType
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerConfig
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerCompressQuality
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerLanguage
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerListener
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerMediaFilter
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerStyle
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerTheme
import io.trtc.tuikit.atomicx.albumpicker.AlbumPickerView

private const val TAG = "UTS-AlbumPickerManager: "

/**
 * Session-specific callbacks storage
 */
private data class SessionCallbacks(
    val onPickConfirm: (String) -> Unit,
    val onMediaProcessing: (String) -> Unit,
    val onMediaProcessed: () -> Unit,
    val onCancel: () -> Unit
)

class AlbumPickerManager private constructor() {

    companion object {
        val shared = AlbumPickerManager()
        private const val OVERLAY_TAG = "AlbumPickerOverlay"
        private const val ANIM_DURATION = 300L
    }

    private var overlayContainer: FrameLayout? = null
    private var pickerView: AlbumPickerView? = null
    private var hostActivity: Activity? = null
    private val gson = Gson()

    /**
     * CRITICAL FIX: Per-session callbacks instead of single instance
     * Maps: sessionId -> SessionCallbacks
     * This prevents callback overwrites when multiple pickers are open simultaneously
     */
    private val sessionCallbacks = mutableMapOf<String, SessionCallbacks>()
    private var currentSessionId: String? = null

    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Show album picker with session isolation
     * @param sessionId - Unique identifier for this picker session
     */
    fun show(
        configJSON: String,
        themeJSON: String,
        sessionId: String,
        activity: Activity,
        onPickConfirm: (String) -> Unit,
        onMediaProcessing: (String) -> Unit,
        onMediaProcessed: () -> Unit,
        onCancel: () -> Unit
    ) {
        console.log("$TAG show called with sessionId: $sessionId")
        Log.i(TAG, "show called with sessionId: $sessionId")

        // Store callbacks with session ID
        this.sessionCallbacks[sessionId] = SessionCallbacks(
            onPickConfirm,
            onMediaProcessing,
            onMediaProcessed,
            onCancel
        )
        this.currentSessionId = sessionId
        this.hostActivity = activity

        // All UI operations must run on the main thread
        mainHandler.post {
            // Remove existing overlay if any
            removeOverlay()

            val decorView = activity.window.decorView as? ViewGroup ?: run {
                console.error("$TAG decorView is null")
                return@post
            }

            // Create a fullscreen overlay container that intercepts BACK key
            // PandoraEntryActivity doesn't forward onBackPressed to OnBackPressedDispatcher,
            // so we intercept the key event here and dispatch it manually.
            val container = object : FrameLayout(activity) {
                override fun dispatchKeyEvent(event: KeyEvent): Boolean {
                    if (event.keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
                        Log.i(TAG, "BACK key intercepted, forwarding to OnBackPressedDispatcher")
                        (activity as? ComponentActivity)?.onBackPressedDispatcher?.onBackPressed()
                        return true
                    }
                    return super.dispatchKeyEvent(event)
                }
            }
            container.isFocusableInTouchMode = true
            container.requestFocus()
            container.tag = OVERLAY_TAG
            container.setBackgroundColor(Color.BLACK)
            container.layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )

            // Propagate lifecycle from Activity's content view so that
            // findViewTreeLifecycleOwner() works inside AlbumPickerView
            val contentView = decorView.findViewById<View>(android.R.id.content)
            val lifecycleOwner = contentView?.findViewTreeLifecycleOwner()
            val savedStateOwner = contentView?.findViewTreeSavedStateRegistryOwner()
            if (lifecycleOwner != null) {
                container.setViewTreeLifecycleOwner(lifecycleOwner)
            }
            if (savedStateOwner != null) {
                container.setViewTreeSavedStateRegistryOwner(savedStateOwner)
            }

            // Create and configure the picker view
            val config = parseConfig(configJSON)
            val theme = parseTheme(themeJSON)

            val picker = AlbumPickerView(activity)

            val listener = object : AlbumPickerListener {
                override fun onPickConfirm(pickedAlbumMedias: List<AlbumMedia>, textMessage: String?) {
                    console.log("$TAG onPickConfirm, count: ${pickedAlbumMedias.size}")

                    val callbacks = sessionCallbacks[sessionId]
                    if (callbacks == null) {
                        console.error("$TAG onPickConfirm: sessionId not found")
                        return
                    }

                    mainHandler.post {
                        animateRemoveOverlay(null)
                    }

                    val serialized = pickedAlbumMedias.map { serializeAlbumMedia(it) }
                    val payload = mapOf(
                        "pickedAlbumMedias" to serialized,
                        "textMessage" to (textMessage ?: "")
                    )
                    val jsonStr = gson.toJson(payload)

                    callbacks.onPickConfirm(jsonStr)
                }

                override fun onMediaProcessing(albumMedia: AlbumMedia, progress: Float, error: Boolean) {
                    
                    val callbacks = sessionCallbacks[sessionId]
                    if (callbacks == null) {
                        console.error("$TAG onMediaProcessing: sessionId not found")
                        return
                    }
                    
                    val payload = mapOf(
                        "albumMedia" to serializeAlbumMedia(albumMedia),
                        "progress" to progress.toDouble(),
                        "error" to error
                    )
                    callbacks.onMediaProcessing(gson.toJson(payload))
                }

                override fun onMediaProcessed() {
                    console.log("$TAG onMediaProcessed")
                    
                    val callbacks = sessionCallbacks[sessionId]
                    if (callbacks == null) {
                        console.error("$TAG onMediaProcessed: sessionId not found")
                        return
                    }
                    
                    callbacks.onMediaProcessed()
                    cleanup(sessionId)
                }

                override fun onCancel() {
                    console.log("$TAG onCancel")

                    val callbacks = sessionCallbacks[sessionId]
                    if (callbacks == null) {
                        console.error("$TAG onCancel: sessionId not found")
                        return
                    }

                    mainHandler.post {
                        animateRemoveOverlay {
                            callbacks.onCancel()
                            cleanup(sessionId)
                        }
                    }
                }
            }

            picker.initialize(config, theme, listener)

            val pickerLayoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            container.addView(picker, pickerLayoutParams)

            // Add overlay to decorView on top of everything
            decorView.addView(container)

            // Slide-up entrance animation
            container.post {
                container.translationY = container.height.toFloat()
                container.animate()
                    .translationY(0f)
                    .setDuration(ANIM_DURATION)
                    .setInterpolator(DecelerateInterpolator())
                    .start()
            }

            this.overlayContainer = container
            this.pickerView = picker

            console.log("$TAG show: overlay added to decorView, sessionId: $sessionId")
            Log.i(TAG, "show: overlay added to decorView, sessionId: $sessionId")
        }
    }

    fun isShowing(): Boolean = overlayContainer != null

    fun handleBackPressed() {
        val activity = hostActivity ?: return
        (activity as? ComponentActivity)?.onBackPressedDispatcher?.onBackPressed()
    }

    private fun removeOverlay() {
        try {
            val container = overlayContainer ?: return
            (container.parent as? ViewGroup)?.removeView(container)
        } catch (e: Exception) {
            Log.e(TAG, "removeOverlay error", e)
        }
        overlayContainer = null
        pickerView = null
    }

    /**
     * Animate the overlay sliding down, then remove it and run onDone.
     * If no overlay exists, onDone runs immediately.
     */
    private fun animateRemoveOverlay(onDone: (() -> Unit)? = null) {
        val container = overlayContainer
        if (container == null) {
            onDone?.invoke()
            return
        }
        container.animate()
            .translationY(container.height.toFloat())
            .setDuration(ANIM_DURATION)
            .setInterpolator(DecelerateInterpolator())
            .setListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    removeOverlay()
                    onDone?.invoke()
                }
            })
            .start()
    }

    /**
     * Clean up a specific session
     */
    private fun cleanup(sessionId: String) {
        sessionCallbacks.remove(sessionId)
        if (currentSessionId == sessionId) {
            removeOverlay()
            currentSessionId = null
        }
        console.log("$TAG cleanup done for sessionId: $sessionId")
        Log.i(TAG, "cleanup done for sessionId: $sessionId")
    }

    // region JSON Parsing

    private fun parseConfig(jsonStr: String): AlbumPickerConfig {
        val config = AlbumPickerConfig()
        try {
            val map = gson.fromJson<MutableMap<String, Any?>>(jsonStr, MutableMap::class.java) ?: return config

            (map["pickMode"] as? Number)?.toInt()?.let { pickMode ->
                config.mediaFilter = when (pickMode) {
                    0 -> AlbumPickerMediaFilter.IMAGE_ONLY
                    1 -> AlbumPickerMediaFilter.VIDEO_ONLY
                    2 -> AlbumPickerMediaFilter.ALL
                    else -> AlbumPickerMediaFilter.ALL
                }
            }

            (map["maxCount"] as? Number)?.toInt()?.let {
                config.maxSelectionCount = it
            }

            (map["gridCount"] as? Number)?.toInt()?.let {
                config.itemsPerRow = it
            }

            (map["showsCameraItem"] as? Boolean)?.let {
                config.showsCameraItem = it
            }

            (map["style"] as? Number)?.toInt()?.let { style ->
                config.style = when (style) {
                    0 -> AlbumPickerStyle.LIKE_WECHAT
                    1 -> AlbumPickerStyle.LIKE_WHATSAPP
                    else -> AlbumPickerStyle.LIKE_WECHAT
                }
            }

            (map["language"] as? Number)?.toInt()?.let { lang ->
                config.language = when (lang) {
                    0 -> AlbumPickerLanguage.SYSTEM
                    1 -> AlbumPickerLanguage.EN
                    2 -> AlbumPickerLanguage.ZH_HANS
                    3 -> AlbumPickerLanguage.ZH_HANT
                    4 -> AlbumPickerLanguage.AR
                    else -> AlbumPickerLanguage.SYSTEM
                }
            }

            (map["compressQuality"] as? Number)?.toInt()?.let { quality ->
                config.compressQuality = when (quality) {
                    0 -> AlbumPickerCompressQuality.STANDARD
                    1 -> AlbumPickerCompressQuality.HIGH
                    else -> AlbumPickerCompressQuality.STANDARD
                }
            }

            (map["maxVideoDurationInSeconds"] as? Number)?.toInt()?.let {
                config.maxVideoDurationInSeconds = it
            }

            (map["maxOutputFileSizeInMB"] as? Number)?.toInt()?.let {
                config.maxOutputFileSizeInMB = it
            }
        } catch (e: Exception) {
            console.error("$TAG parseConfig failed: ${e.message}")
            Log.e(TAG, "parseConfig failed", e)
        }
        return config
    }

    private fun parseTheme(jsonStr: String): AlbumPickerTheme {
        val theme = AlbumPickerTheme()
        try {
            val map = gson.fromJson<MutableMap<String, Any?>>(jsonStr, MutableMap::class.java) ?: return theme

            theme.currentPrimaryColor = parseColor(map["primaryColor"])
            theme.backgroundColor = parseColor(map["backgroundColor"])
            theme.backgroundColorSecondary = parseColor(map["backgroundColorSecondary"])
            theme.textColor = parseColor(map["textColor"])
            theme.textColorSecondary = parseColor(map["textColorSecondary"])

            (map["bigFontSize"] as? Number)?.toFloat()?.let { theme.bigFontSize = it }
            (map["normalFontSize"] as? Number)?.toFloat()?.let { theme.normalFontSize = it }
            (map["smallFontSize"] as? Number)?.toFloat()?.let { theme.smallFontSize = it }
            (map["bigRadius"] as? Number)?.toInt()?.let { theme.bigRadius = it }
            (map["normalRadius"] as? Number)?.toInt()?.let { theme.normalRadius = it }
            (map["smallRadius"] as? Number)?.toInt()?.let { theme.smallRadius = it }
        } catch (e: Exception) {
            console.error("$TAG parseTheme failed: ${e.message}")
            Log.e(TAG, "parseTheme failed", e)
        }
        return theme
    }

    private fun parseColor(value: Any?): Int? {
        val hexStr = value as? String ?: return null
        val hex = hexStr.replace("0x", "").replace("0X", "").replace("#", "")
        if (hex.length != 8) return null
        return try {
            java.lang.Long.parseLong(hex, 16).toInt()
        } catch (e: Exception) {
            null
        }
    }

    private fun serializeAlbumMedia(media: AlbumMedia): Map<String, Any?> {
        return mapOf(
            "id" to media.id.toLong(),
            "mediaType" to if (media.mediaType == AlbumMediaType.VIDEO) 1 else 0,
            "mediaPath" to (media.mediaPath ?: ""),
            "fileExtension" to "",
            "fileSize" to 0,
            "videoThumbnailPath" to media.videoThumbnailPath,
            "duration" to media.duration,
        )
    }

    // endregion
}
